File: PASCAL/conversoes.pas
unit conversoes;

{$mode objfpc}{$H+}

interface

uses
  SysUtils, Classes, Generics.Collections, tipos, automato;

function AFNParaAFD(const AFN: TAutomato): TAutomato;
function AFDParaAFN(const AFD: TAutomato): TAutomato;
function AFNParaAFNLambda(const AFN: TAutomato): TAutomato;
function AFDParaAFNLambda(const AFD: TAutomato): TAutomato;

implementation

type
  TEstadoAFD = class
    Estados: TStringSet;
    Nome: String;
    constructor Create;
    destructor Destroy; override;
  end;

constructor TEstadoAFD.Create;
begin
  Estados := TStringSet.Create;
end;

destructor TEstadoAFD.Destroy;
begin
  Estados.Free;
  inherited;
end;

function ChaveTransicao(const Estado, Simbolo: String): String;
begin
  Result := Estado + '|' + Simbolo;
end;

function ConjuntoParaNome(Estados: TStringSet): String;
var
  Lista: TStringList;
  S: String;
  I: Integer;
begin
  if Estados.Count = 0 then
    Exit('{}');
  
  Lista := TStringList.Create;
  try
    for S in Estados do
      Lista.Add(S);
    Lista.Sort;
    
    Result := '{';
    for I := 0 to Lista.Count - 1 do
    begin
      Result := Result + Lista[I];
      if I < Lista.Count - 1 then
        Result := Result + ',';
    end;
    Result := Result + '}';
  finally
    Lista.Free;
  end;
end;

function AFNParaAFD(const AFN: TAutomato): TAutomato;
var
  TabelaAFN: TTabela;
  EstadoInicialAFD: TStringSet;
  Fila: specialize TList<TEstadoAFD>;
  Processados: specialize TList<String>;
  EstadoAtual: TEstadoAFD;
  NomeAtual: String;
  Simbolo, EstadoAFN, Proximo: String;
  ProximoEstado: TStringSet;
  Destinos: TStringSet;
  Chave: String;
  Trans: TTransicao;
  Temp: TEstadoAFD;
begin
  Result := CriarAutomato;
  TabelaAFN := ConstruirTabelaTransicoes(AFN);
  EstadoInicialAFD := FechoLambda(AFN.EstadosIniciais, TabelaAFN);
  
  Fila := specialize TList<TEstadoAFD>.Create;
  Processados := specialize TList<String>.Create;
  
  try
    Temp := TEstadoAFD.Create;
    Temp.Estados.Free;
    Temp.Estados := EstadoInicialAFD;
    Temp.Nome := ConjuntoParaNome(EstadoInicialAFD);
    Fila.Add(Temp);
    
    while Fila.Count > 0 do
    begin
      EstadoAtual := Fila[0];
      Fila.Delete(0);
      NomeAtual := EstadoAtual.Nome;
      
      if Processados.IndexOf(NomeAtual) >= 0 then
      begin
        EstadoAtual.Free;
        Continue;
      end;
      
      Processados.Add(NomeAtual);
      AdicionarAoConjunto(Result.Estados, NomeAtual);
      
      for Proximo in EstadoAtual.Estados do
        if ConjuntoContem(AFN.EstadosFinais, Proximo) then
        begin
          AdicionarAoConjunto(Result.EstadosFinais, NomeAtual);
          Break;
        end;
      
      for Simbolo in AFN.Alfabeto do
      begin
        if Simbolo = '&' then Continue;
        
        ProximoEstado := TStringSet.Create;
        
        for EstadoAFN in EstadoAtual.Estados do
        begin
          Chave := ChaveTransicao(EstadoAFN, Simbolo);
          if TabelaAFN.TryGetValue(Chave, Destinos) then
            for Proximo in Destinos do
              AdicionarAoConjunto(ProximoEstado, Proximo);
        end;
        
        if ProximoEstado.Count > 0 then
        begin
          Temp := TEstadoAFD.Create;
          Temp.Estados.Free;
          Temp.Estados := FechoLambda(ProximoEstado, TabelaAFN);
          Temp.Nome := ConjuntoParaNome(Temp.Estados);
          
          Trans.Origem := NomeAtual;
          Trans.Destino := Temp.Nome;
          Trans.Simbolo := Simbolo;
          Result.Transicoes.Add(Trans);
          
          if Processados.IndexOf(Temp.Nome) < 0 then
            Fila.Add(Temp)
          else
            Temp.Free;
        end;
        
        ProximoEstado.Free;
      end;
      
      EstadoAtual.Free;
    end;
    
    for Simbolo in AFN.Alfabeto do
      if Simbolo <> '&' then
        AdicionarAoConjunto(Result.Alfabeto, Simbolo);
    
    NomeAtual := ConjuntoParaNome(EstadoInicialAFD);
    AdicionarAoConjunto(Result.EstadosIniciais, NomeAtual);
    
  finally
    for Temp in Fila do
      Temp.Free;
    Fila.Free;
    Processados.Free;
    LiberarTabela(TabelaAFN);
  end;
end;

function AFDParaAFN(const AFD: TAutomato): TAutomato;
var
  S: String;
  Trans: TTransicao;
begin
  Result := CriarAutomato;
  
  for S in AFD.Alfabeto do
    Result.Alfabeto.Add(S);
  for S in AFD.Estados do
    Result.Estados.Add(S);
  for S in AFD.EstadosIniciais do
    Result.EstadosIniciais.Add(S);
  for S in AFD.EstadosFinais do
    Result.EstadosFinais.Add(S);
  for Trans in AFD.Transicoes do
    Result.Transicoes.Add(Trans);
end;

function AFNParaAFNLambda(const AFN: TAutomato): TAutomato;
var
  S, EstadoInicial, Simbolo, Destino: String;
  Trans, NovoTrans: TTransicao;
  Tabela: TTabela;
  Destinos: TStringSet;
  Chave: String;
  Existe: Boolean;
  T: TTransicao;
begin
  Result := CriarAutomato;
  
  for S in AFN.Alfabeto do
    Result.Alfabeto.Add(S);
  for S in AFN.Estados do
    Result.Estados.Add(S);
  for S in AFN.EstadosIniciais do
    Result.EstadosIniciais.Add(S);
  for S in AFN.EstadosFinais do
    Result.EstadosFinais.Add(S);
  for Trans in AFN.Transicoes do
    Result.Transicoes.Add(Trans);
  
  if not ConjuntoContem(Result.Alfabeto, '&') then
    AdicionarAoConjunto(Result.Alfabeto, '&');
  
  Tabela := ConstruirTabelaTransicoes(AFN);
  
  try
    for EstadoInicial in AFN.EstadosIniciais do
      for Simbolo in AFN.Alfabeto do
      begin
        Chave := ChaveTransicao(EstadoInicial, Simbolo);
        if Tabela.TryGetValue(Chave, Destinos) then
          for Destino in Destinos do
          begin
            Existe := False;
            for T in Result.Transicoes do
              if (T.Origem = EstadoInicial) and (T.Destino = Destino) and (T.Simbolo = '&') then
              begin
                Existe := True;
                Break;
              end;
            
            if not Existe then
            begin
              NovoTrans.Origem := EstadoInicial;
              NovoTrans.Destino := Destino;
              NovoTrans.Simbolo := '&';
              Result.Transicoes.Add(NovoTrans);
            end;
          end;
      end;
  finally
    LiberarTabela(Tabela);
  end;
end;

function AFDParaAFNLambda(const AFD: TAutomato): TAutomato;
var
  AFN: TAutomato;
begin
  AFN := AFDParaAFN(AFD);
  try
    Result := AFNParaAFNLambda(AFN);
  finally
    LiberarAutomato(AFN);
  end;
end;

end.

File: src/conversoes.py
from typing import List, Set, Dict
from automato import Automato

def afn_para_afd(afn: Automato) -> Automato:
    tabela_afn = afn.construir_tabela_transicoes()
    estado_inicial_afd = frozenset(afn.fecho_lambda(afn.estados_iniciais, tabela_afn))
    
    estados_afd = set()
    transicoes_afd = []
    estados_finais_afd = set()
    
    fila = [estado_inicial_afd]
    processados = set()
    
    while fila:
        estado_atual = fila.pop(0)
        
        if estado_atual in processados:
            continue
        
        processados.add(estado_atual)
        estados_afd.add(estado_atual)
        
        if estado_atual & afn.estados_finais:
            estados_finais_afd.add(estado_atual)
        
        for simbolo in afn.alfabeto:
            if simbolo == '&':
                continue
            
            proximo_estado = set()
            for estado_afn in estado_atual:
                if (estado_afn, simbolo) in tabela_afn:
                    proximo_estado.update(tabela_afn[(estado_afn, simbolo)])
            
            proximo_estado = afn.fecho_lambda(proximo_estado, tabela_afn)
            
            if proximo_estado:
                proximo_estado = frozenset(proximo_estado)
                
                transicoes_afd.append([
                    converter_estado_afd_para_nome(estado_atual),
                    converter_estado_afd_para_nome(proximo_estado),
                    simbolo
                ])
                
                if proximo_estado not in processados:
                    fila.append(proximo_estado)
    
    estados_afd_nomes = {converter_estado_afd_para_nome(e) for e in estados_afd}
    estados_finais_afd_nomes = {converter_estado_afd_para_nome(e) for e in estados_finais_afd}
    estado_inicial_afd_nome = converter_estado_afd_para_nome(estado_inicial_afd)
    
    alfabeto_afd = afn.alfabeto - {'&'}
    
    return Automato(
        alfabeto=list(alfabeto_afd),
        estados=list(estados_afd_nomes),
        estados_iniciais=[estado_inicial_afd_nome],
        estados_finais=list(estados_finais_afd_nomes),
        transicoes=transicoes_afd
    )


def converter_estado_afd_para_nome(estado_frozenset) -> str:
    if not estado_frozenset:
        return '{}'
    return '{' + ','.join(sorted(estado_frozenset)) + '}'


def afd_para_afn(afd: Automato) -> Automato:
    return Automato(
        alfabeto=list(afd.alfabeto),
        estados=list(afd.estados),
        estados_iniciais=list(afd.estados_iniciais),
        estados_finais=list(afd.estados_finais),
        transicoes=afd.transicoes.copy()
    )


def afn_para_afn_lambda(afn: Automato) -> Automato:
    novos_estados = list(afn.estados)
    novas_transicoes = afn.transicoes.copy()
    novo_alfabeto = list(afn.alfabeto)
    
    if '&' not in novo_alfabeto:
        novo_alfabeto.append('&')
    
    tabela = afn.construir_tabela_transicoes()
    
    for estado_inicial in afn.estados_iniciais:
        for simbolo in afn.alfabeto:
            if (estado_inicial, simbolo) in tabela:
                for destino in tabela[(estado_inicial, simbolo)]:
                    if [estado_inicial, destino, '&'] not in novas_transicoes:
                        novas_transicoes.append([estado_inicial, destino, '&'])
    
    return Automato(
        alfabeto=novo_alfabeto,
        estados=novos_estados,
        estados_iniciais=list(afn.estados_iniciais),
        estados_finais=list(afn.estados_finais),
        transicoes=novas_transicoes
    )


def afd_para_afn_lambda(afd: Automato) -> Automato:
    afn = afd_para_afn(afd)
    return afn_para_afn_lambda(afn)
